### ISSUE TITLE
Weight initializers return host arrays

### ISSUE THREAD

**Issue by jonasrauber** [google/jax#245]
Weight initializers in stax, e.g. `glorot()`, return standard numpy arrays. Wouldn't it be better to move them to the GPU (is `jax.device_put()` the right way?) to avoid that they need to get copied to the GPU all the time?
Also: is there an easy way to see if an array is on the GPU or not or is type-checking the recommended way?

**Comment by mattjj** [comment]
You're right: it would be better to initialize weights on the device, rather than transferring them on the first update. Luckily they are only copied on the first update step, so there's no need for an explicit `jax.device_put()`. But still it's best not to have to do that transfer, even the first time.

Regarding how to check if an array is backed by device memory, type-checking is indeed the current way to do it. Something like this would work:

```python
from jax.interpreters import xla

def is_device_array(x):
  return isinstance(x, xla.DeviceArray)
```

Maybe we should add that as a convenience method to api.py. What do you think?

**Comment by jekbradbury** [comment]
This was fixed in #566.
